### PR TITLE
Refactor: split unit tests by API resources

### PR DESCRIPTION
**Pull request by nejch**
I've been thinking about how to make testing more accessible for contributors and maybe splitting them by resources in the Gitlab API docs would be best because that's how most people interact with this anyway. If there's a new API resource it likely gets a new page so this would kind of make it a 1-to-1 for new contributions.

Ideas:
- [x] Split unit tests by API resources
- [x] Convert unittest classes into modules? (disclaimer: really biased here against classes :P) 
- [x] ~~Figure out a better way for fixtures & mocking copy/pasting~~

### CODE REVIEW

File: gitlab/__init__.py [hunk 1]
```diff
@@ -321,7 +311,7 @@ def set_license(self, license, **kwargs):
     def _construct_url(self, id_, obj, parameters, action=None):
         if "next_url" in parameters:
             return parameters["next_url"]
-        args = _sanitize(parameters)
+        args = utils.sanitize_parameters(parameters)
```
Comment (max-wittig) [inline gitlab/__init__.py:314]: nice change. Get rid of custom functions!

File: gitlab/tests/objects/test_application.py [hunk 2]
```diff
@@ -1,120 +1,131 @@
...
-            path="/api/v4/application/appearance",
-            method="put",
+def test_update_application_appearance(gl):
+    @urlmatch(
+        scheme="http",
+        netloc="localhost",
+        path="/api/v4/application/appearance",
+        method="put",
+    )
+    def resp_update_appearance(url, request):
```
[the remark this answers is not in the document]
Comment (nejch) [reply gitlab/tests/objects/test_application.py:111]: Oops, probably some leftovers from the copy/paste. Late-night commits :grin: I've taken a few more duplicate responses out of other tests now so they can be shared where possible.

File: gitlab/tests/objects/test_users.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+    return response(200, content, headers, None, 5, request)
+
+
+@urlmatch(
+    scheme="http", netloc="localhost", path="/api/v4/users/1/memberships", method="get",
+)
+def resp_get_user_memberships(url, request):
+    headers = {"content-type": "application/json"}
+    content = """[
+                  {
```
[the remark this answers is not in the document]
Comment (nejch) [reply gitlab/tests/objects/test_users.py:28]: Thanks! I've made it look more like the mirrors example response now.

**Comment by nejch** [comment]
> Looks good overall so far!
> 
> I'm also biased against classes in python. See also the last sentence of the Zen of Python

:grin: that's great I'm glad, I was afraid these changes would be too much. I've only left the 2 classes in `test_base.py` for now as it's a small file and didn't seem to make sense to split it yet.

I realized I might be getting myself into a continuous rebase party as new PR's with tests in the old files are merged, so I've cleaned it up a bit more and I think I want to look at fixtures in a later PR when I have more time, if this is in an OK shape?

**Comment by max-wittig** [comment]
@nejch Sorry totally missed this. Does one only see assignments and not reviewers in the feed?

File: gitlab/tests/mixins/test_object_mixins_attributes.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
+# -*- coding: utf-8 -*-
```
Comment (max-wittig) [inline gitlab/tests/mixins/test_object_mixins_attributes.py:1]: Where does this license header come from?

**Comment by max-wittig** [comment]
@nejch I would love to merge this. Could you please rebase to latest master?

File: gitlab/tests/mixins/test_object_mixins_attributes.py (hunk 4 above)
Comment (nejch) [reply gitlab/tests/mixins/test_object_mixins_attributes.py:1]: Oops I must have left this when splitting the files from the original mixin tests. I've removed it now but I can also add new headers to all the new files if needed.

**Comment by nejch** [comment]
> 
> 
> @nejch I would love to merge this. Could you please rebase to latest master?

Sorry @max-wittig I'm back from the dead/vacation/other distractions now 😁 I've rebased and rewritten the tests using responses now so it's more consistent with your runners example. Could you have another look to see if this is how you imagined `responses`-based tests?

Not sure about notifications, there's a separate Notification setting for PR Reviews so maybe you had them turned off?

**Comment by max-wittig** [comment]
We actually have less tests with this MR. I guess some got lost. Could yout take a look? @nejch 

This MR: https://travis-ci.org/github/python-gitlab/python-gitlab/jobs/720445167
Master: https://travis-ci.org/github/python-gitlab/python-gitlab/jobs/720460657

**Comment by nejch** [comment]
> We actually have less tests with this MR. I guess some got lost. Could yout take a look? @nejch
> 
> 155 vs. 158
> 
> This MR: https://travis-ci.org/github/python-gitlab/python-gitlab/jobs/720445167
> Master: https://travis-ci.org/github/python-gitlab/python-gitlab/jobs/720460657

Oof, I was weirded out by this until I realized that on master, 2 test cases (in `test_groups.py`) were run 3 times. They're duplicates because I had accidentally put them in a unittest base class (TestGroup) inherited by other test classes back when I was adding group export/import tests. :facepalm: 
With pytest they're run only once, hence the lower number. If you exclude `test_groups.py` the number should be higher. But I checked all cases collected in master vs. this PR again and it should be OK otherwise.

With `tox -e py38 -- -v`:

master:
```
gitlab/tests/objects/test_groups.py::TestGroup::test_create_group PASSED [ 75%]
gitlab/tests/objects/test_groups.py::TestGroup::test_get_group PASSED    [ 75%]
gitlab/tests/objects/test_groups.py::TestGroupExport::test_create_group PASSED [ 76%]
gitlab/tests/objects/test_groups.py::TestGroupExport::test_create_group_export PASSED [ 76%]
gitlab/tests/objects/test_groups.py::TestGroupExport::test_download_group_export PASSED [ 77%]
gitlab/tests/objects/test_groups.py::TestGroupExport::test_get_group PASSED [ 78%]
gitlab/tests/objects/test_groups.py::TestGroupExport::test_refresh_group_export_status SKIPPED [ 78%]
gitlab/tests/objects/test_groups.py::TestGroupImport::test_create_group PASSED [ 79%]
gitlab/tests/objects/test_groups.py::TestGroupImport::test_get_group PASSED [ 80%]
gitlab/tests/objects/test_groups.py::TestGroupImport::test_import_group PASSED [ 80%]
gitlab/tests/objects/test_groups.py::TestGroupImport::test_refresh_group_import_status SKIPPED [ 81%]
```

refactor:
```
gitlab/tests/objects/test_groups.py::test_get_group PASSED               [ 56%]
gitlab/tests/objects/test_groups.py::test_create_group PASSED            [ 56%]
gitlab/tests/objects/test_groups.py::test_create_group_export PASSED     [ 57%]
gitlab/tests/objects/test_groups.py::test_refresh_group_export_status SKIPPED [ 57%]
gitlab/tests/objects/test_groups.py::test_download_group_export PASSED   [ 58%]
gitlab/tests/objects/test_groups.py::test_import_group PASSED            [ 58%]
gitlab/tests/objects/test_groups.py::test_refresh_group_import_status SKIPPED [ 59%]
```

**Comment by max-wittig** [comment]
Thank you very much! Very nice!

**Pull request merged**